aten/src/ATen/native/cuda/Loss.cu

```diff
@@ -153,4 +154,290 @@ Tensor& binary_cross_entropy_backward_out_cuda(const Tensor& grad, const Tensor&
+  // NOLINTNEXTLINE(cppcoreguidelines-init-variables)
+  int i, t;
+  // NOLINTNEXTLINE(cppcoreguidelines-init-variables)
+  scalar_t cur_weight;
+
+  sh_inputs[threadIdx.x] = static_cast<accscalar_t>(0);
+  acc_weight[threadIdx.x] = static_cast<accscalar_t>(0);
+  for (i = threadIdx.x; i < nframe; i += NLL_LOSS_THREADS) {
+    t = target[i];
+    if (t != static_cast<int>(ignore_index)) {
+      CUDA_KERNEL_ASSERT(t >= 0 && t < n_classes);
+      cur_weight = weights != nullptr ? weights[t] : static_cast<scalar_t>(1);
+      sh_inputs[threadIdx.x] -= input[i * ndim + t] * cur_weight;
+      acc_weight[threadIdx.x] += cur_weight;
+    }
+  }
+
+  __syncthreads();
+
+  if (threadIdx.x == 0) {
+    *output = *total_weight = static_cast<scalar_t>(0);
+    accscalar_t output_acc = 0;
+    accscalar_t total_weight_acc = 0;
+    for (i = 0; i < NLL_LOSS_THREADS; ++i) {
+      output_acc += sh_inputs[i];
+      total_weight_acc += acc_weight[i];
+    }
+    *total_weight = static_cast<scalar_t>(total_weight_acc);
+    *output = static_cast<scalar_t>(output_acc);
+    if (size_average) {
+      if (nframe == 0) {
+        // Mean reduction on empty tensors produces NaN
+        *output = std::numeric_limits<double>::quiet_NaN();
+      }
+      if (*total_weight != 0) {
+        *output = static_cast<scalar_t>(output_acc / total_weight_acc);
+      }
+    }
+  }
+}
+
+} // namespace
+
+void nll_loss_forward_out_cuda_template(
+    Tensor& output,
+    Tensor& total_weight,
+    const Tensor& input,
+    const Tensor& target,
+    const c10::optional<Tensor>& weight_opt,
+    int64_t reduction,
+    int64_t ignore_index) {
+  // See [Note: hacky wrapper removal for optional tensor]
+  c10::MaybeOwned<Tensor> weight_maybe_owned =
+      at::borrow_from_optional_tensor(weight_opt);
+  const Tensor& weight = *weight_maybe_owned;
+
+  TORCH_CHECK(target.dim() <= 1, "multi-target not supported");
+
+  int64_t n_classes = input.size(-1);
+  int64_t n_dims = input.dim();
+
+  TensorArg output_arg{output, "output", 1};
+  TensorArg total_weight_arg{total_weight, "total_weight", 2};
+  TensorArg input_arg{input, "input", 3};
+  TensorArg target_arg{target, "target", 4};
+
+  if (weight.defined()) {
+    TensorArg weight_arg{weight, "weight", 5};
+    checkAllSameGPU(
+        "nll_loss_forward_out_cuda_template",
+        {output_arg, total_weight_arg, input_arg, target_arg, weight_arg});
+  } else {
+    checkAllSameGPU(
+        "nll_loss_forward_out_cuda_template",
+        {output_arg, total_weight_arg, input_arg, target_arg});
+  }
+
+  TORCH_CHECK(n_dims > 0 && n_dims <= 2, "input tensor should be 1D or 2D");
+  int64_t batch_size = n_dims == 1 ? 1 : input.size(0);
+  int64_t num_targets = target.size(0);
+  TORCH_CHECK(
+      batch_size == num_targets,
+      "size mismatch (got input: ",
+      input.sizes(),
+      ", target: ",
+      target.sizes(),
+      ")")
+
+  TORCH_CHECK(
+      !weight.defined() || weight.numel() == n_classes,
```

Comment:
should you check weight.dim() here (and also add a test case for incorrectly sized/dimensioned weights?)

Reply:
I added more tests for input validation and also made a slight adjustment to the CPU version to also check for `weight.dim`.